### PR TITLE
Added auditing, alter_view, semantic predicate for version 12 of Oracle, analyze_table, associate statistics, reworked keywords

### PR DESCRIPTION
**Pull request by madams51703**

### CODE REVIEW

File: plsql/PlSqlParser.g4 [hunk 1]
```diff
@@ -1,4 +1,4 @@
- /**
+ /**CTIO
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: What is it?

File: plsql/PlSqlParser.g4 [hunk 2]
```diff
@@ -21,7 +21,7 @@
 parser grammar PlSqlParser;
 
 options { tokenVocab=PlSqlLexer; }
-
+@members {boolean version12=true;}
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: Is it possible to move runtime dependent code to super classes? See [JavaScript grammar](https://github.com/antlr/grammars-v4/tree/master/javascript) for example.

File: plsql/PlSqlParser.g4 [hunk 3]
```diff
@@ -899,6 +905,268 @@ container_data_clause
...
+                )
+    | SUBPARTITION ( '(' subpartition_name ')'
+                   | FOR '(' (','? subpartition_key_value)+ ')'
+                   )
+    ;
+
+validation_clauses
+    : VALIDATE REF UPDATE (SET DANGLING TO NULL)?
+    | VALIDATE STRUCTURE 
+        ( CASCADE  FAST
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: Extra space.

File: plsql/PlSqlParser.g4 [hunk 4]
```diff
@@ -899,6 +905,268 @@ container_data_clause
...
+    ;
+
+storage_table_clause
+    : WITH (SYSTEM | USER) MANAGED STORAGE TABLES
+    ;
+
+// https://docs.oracle.com/database/121/SQLRF/statements_4008.htm#SQLRF56110
+unified_auditing
+    : {version12}? 
+      AUDIT (POLICY policy_name ((BY|EXCEPT) (','? audit_user)+ )?
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: Add spaces here, please.

File: plsql/PlSqlParser.g4 [hunk 5]
```diff
@@ -899,6 +905,268 @@ container_data_clause
...
+audit_direct_path
+    : {version12}? DIRECT_PATH auditing_by_clause
+    ;
+
+audit_container_clause
+    : {version12}? (CONTAINER EQUALS_OP (CURRENT | ALL))
+    ;
+
+audit_operation_clause
+    : ( (','? (sql_statement_shortcut | ALL | ALL STATEMENTS) )+
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: `ALL | ALL STATEMENTS` -> `ALL? STATEMENTS`.

File: plsql/PlSqlParser.g4 [hunk 6]
```diff
@@ -899,6 +905,268 @@ container_data_clause
...
+audit_user
+    : regular_id
+    ;
+
+audit_schema_object_clause
+    : ( (','? sql_operation)+
+      | ALL
+      )
+
+      auditing_on_clause
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]: Maybe rewrite it to a single line?

```antlr
audit_schema_object_clause
    : ((','? sql_operation)+ | ALL) auditing_on_clause
    ;
```

File: plsql/PlSqlParser.g4 [hunk 7]
```diff
@@ -899,6 +905,268 @@ container_data_clause
...
+subpartition_key_value
+    : literal
+    ;
+
+//https://docs.oracle.com/cd/E11882_01/server.112/e41084/statements_4006.htm#SQLRF01106
+associate_statistics
+    : ASSOCIATE STATISTICS 
+        WITH ( column_association
+             | function_association
+             )
```
Comment (KvanTTT) [inline plsql/PlSqlParser.g4]:
```antlr
WITH (column_association | function_association)
```

File: plsql/PlSqlParser.g4 (hunk 1 above)
Comment (madams51703) [reply plsql/PlSqlParser.g4]: Garbage, removing it.

File: plsql/PlSqlParser.g4 (hunk 2 above)
Comment (madams51703) [reply plsql/PlSqlParser.g4]: I made the other changes.  I am not yet understanding how to make this change.  I read the example and it is not yet clear to me.  If I can figure it out, I will definitely make this change.

Comment (madams51703) [reply plsql/PlSqlParser.g4]: I tried generating the javascript example and get errors trying to compile the java classes(I am using 4.7.1):

javac *.java
JavaScriptLexer.java:12: error: cannot find symbol
public class JavaScriptLexer extends JavaScriptBaseLexer {
                                     ^
  symbol: class JavaScriptBaseLexer
JavaScriptParser.java:12: error: cannot find symbol
public class JavaScriptParser extends JavaScriptBaseParser {
                                      ^
  symbol: class JavaScriptBaseParser
JavaScriptLexer.java:133: error: method does not override or implement a method from a supertype
	@Override

Comment (KvanTTT) [reply plsql/PlSqlParser.g4]: It seems to me that you forgot to include the superclass files `JavaScriptBaseLexer` and `JavaScriptBaseParser` from java directory.

Comment (KvanTTT) [reply plsql/PlSqlParser.g4]: Ok, I'll fix it by myself later.

**Comment by madams51703** [comment]
Thanks.  I am starting to understand it.  If/When I figure it out I will
fix it.

**Comment by madams51703** [comment]
I just now saw the example superclass for javascript.  I will try to get it
working for my next release.

**Comment by madams51703** [comment]
Ok figured it out.  I was pretty dense about it, but it makes sense now.

**Pull request merged**